enumflags_derive/src/lib.rs

```diff
@@ -181,21 +182,21 @@ fn gen_enumflags(ident: &Ident, item: &DeriveInput, data: &DataEnum) -> TokenStr
                        fmt: &mut #std_path::fmt::Formatter)
                        -> #std_path::fmt::Result {
                     use ::enumflags2::RawBitFlags;
-                    let v:Vec<&str> =
-                        [#((#names :: #variants).bits(),)*]
-                        .iter()
-                        .filter_map(|val|{
-                            let val: #ty = *val as #ty & flags.bits();
-                            match val {
-                                #(#flag_value_names => Some(stringify!(#variants)),)*
-                                _ => None
-                            }
-                        })
-                        .collect();
-                    write!(fmt, "BitFlags<{}>(0b{:b}, [{}]) ",
+                    use #std_path::iter::Iterator;
+                    const VARIANT_NAMES: [&'static str; #variants_len] = [#(#variants_names, )*];
+                    let vals =
+                        VARIANTS.iter().zip(VARIANT_NAMES.iter())
+                        .filter(|&(&val, _)| val as #ty & flags.bits() != 0)
+                        .map(|(_, name)| name);
+                    write!(fmt, "BitFlags<{}>(0b{:b}, [",
                            stringify!(#ident),
-                           flags.bits(),
-                           v.join(", "))
+                           flags.bits())?;
+                    for (i, val) in vals.enumerate() {
+                        write!(fmt, "{}{}",
+                               if i == 0 { "" } else { ", " },
+                               val)?;
+                    }
+                    write!(fmt, "]) ")
```

Comment:
I was wondering that myself, but just kept the original string as-is. I'm assuming debug strings aren't supposed to have trailing space, should I remove it?

(though also on the topic of formatting I might ask why it's `[A, B]` and not `A | B`)

Reply:
This is a very good question. However, I'm not the one who wrote this code in the first place. If I were to guess, the latter option didn't even occur to the author. I won't object if you change this to ` | `. Make sure the output is reasonable for an empty bitflag if you do, though. Perhaps the output in that case could just be `BitFlags<T>(0b0)`? You could even implement this quite neatly: `if let Some(first_flag) = vals.next()`, and then the `.enumerate()` would become unnecessary.

Reply:
Though the elegance is slightly tarnished by `no_implicit_prelude`, sounds good to me!

Reply:
So after implementing this, I'm wondering if having a list of variants names may just be misguided to begin with. If the underlying enum is already `Debug`, why can't the impl Debug in BitFlags just `self.iter().map(|val| write!(fmt, " | {:?}", val))` instead? It even composes better with custom Debug impls as a bonus.

Downsides:
1. Slightly less efficient than just having a `[&str]` list all ready to go
2. You'll get `T::A | T::B` rather than `A | B` if you allow the default `derive(Debug)` to implement it for the enum.

Reply:
This thought also occurred to me, but then I realized that the underlying enum is not always Debug. I don't think anybody is going to be too upset if such a requirement is added for `impl Debug for BitFlags`, though.

Efficiency is arguable. While printing out the bitflag is going to take longer, the code will be smaller because the strings won't be duplicated in the binary (assuming that the underlying enum is also implementing Debug). Frankly, if the Debug impl is a hot spot in your program, you're doing something horribly wrong, and a smaller binary is more valuable in this case.

Reply:
> I don't think anybody is going to be too upset if such a requirement is added for impl Debug for BitFlags, though.

If anything, supporting `Debug` without a `derive(Debug)` on the enum declaration is a bit surprising. Hardly an unwanted surprise sure, but it seems reasonable to expect debug-printing an enum to fail without having defined it to be `Debug`. It also seems maybe desireable to use a custom `Debug` in cases where canon flag names don't match up with rust naming conventions, though maybe that's best left to `Display`, dunno.

> Efficiency is arguable. While printing out the bitflag is going to take longer, the code will be smaller because the strings won't be duplicated in the binary (assuming that the underlying enum is also implementing Debug). Frankly, if the Debug impl is a hot spot in your program, you're doing something horribly wrong, and a smaller binary is more valuable in this case

Presumably the strings will get intern'd so the const array will just be a list of pointers to the same strings Debug uses, so it probably doesn't make things worse, but... yeah it's unnecessary and simplifying codegen and impls trumps having a handy list to speed up an irrelevant operation.